packages/vite/src/node/ssr/ssrModuleLoader.ts

```diff
@@ -174,18 +182,36 @@ async function instantiateModule(
   return Object.freeze(ssrModule)
 }
 
-function nodeRequire(
+// In node@12+ we can use dynamic import to load CJS and ESM
+// Caveat https://github.com/vitejs/vite/pull/5197#discussion_r722583544
+async function nodeImport(
   id: string,
   importer: string | null,
-  root: string,
-  preserveSymlinks: boolean
+  config: ViteDevServer['config'],
 ) {
-  const mod = require(resolve(id, importer, root, preserveSymlinks))
-  const defaultExport = mod.__esModule ? mod.default : mod
-  // rollup-style default import interop for cjs
+  const url = pathToFileURL(resolve(id, importer, config.root, !!config.resolve.preserveSymlinks))
```

Comment:
Two discussion points here:
1. `pathToFileURL` is needed for Windows compat. This is the very first thing that bites anyone writing Node ESM.
2. Using `resolve` here does **not** match Node behavior where import specifiers in ESM **must be** explicit. Node does not implement the same resolution algorithm for ESM and CJS. I'd argue that Vite shouldn't use the CJS resolution algorithm for ESM, because the split was an explicit choice by the Node team to reduce implementation complexity and mental overhead. Unfortunately, I'm not sure that this will work without using `resolve`. Very open to ideas... maybe try `import()` and fallback to `require()`?

Reply:
2. We could use `moduleResolve` from [this package](https://github.com/wooorm/import-meta-resolve#moduleresolvespecifier-parent-conditions).

Reply:
Rollup does typically allow importing without the `.js` extension, so I don't think we'd need to or want to require the `.js` extension. I tend to think it'd be a bit confusing or lead to more problems if Vite behaved differently than Rollup there since Vite is supposed to be an extension of Rollup